File: python/LAHMC.py
import numpy as np

class HMC_state(object):
    """ Holds all the state variables for qNHMC. """

    def __init__(self, X, parent, V=None, EX=None, EV=None, dEdX=None):
        self.parent = parent
        self.X = X
        self.V = V
        if V is None:
            N = self.X.shape[0]
            nbatch = self.X.shape[1]
            self.V = np.random.randn(N, nbatch)
        self.EX = EX
        if EX is None:
            self.update_EX()
        self.EV = EV
        if EV is None:
            self.EV = self.update_EV()
        self.dEdX = dEdX
        if dEdX is None:
            self.update_dEdX()
        self.active_idx = False

    def update_EX():
        if self.active_idx:
            self.EX[:,self.active_idx] = parent.E(self.X[:,self.active_idx])
        else:
            self.EX = parent.E(self.X)
    def update_EV():
        if self.active_idx:
            self.EV[:,self.active_idx] = np.sum(self.V[:,self.active_idx]**2, axis=0).reshape((1,-1))/2.
        else:
            self.EV = np.sum(self.V**2, axis=0).reshape((1,-1))/2.
    def update_dEdX():
        if self.active_idx:
            self.dEdX[:,self.active_idx] = parent.dEdX(self.X[:,self.active_idx])
        else:
            self.dEdX = parent.dEdX(self.X)

    def copy(self):
        Z = HMC_state(self.X.copy(), self.parent, V=self.V.copy(), EX=self.EX.copy(), EV=self.EV.copy(), dEdX=self.dEdX.copy())
        return Z

    def update(self, idx, Z):
        """ replace batch elements idx with state from Z """
        if len(idx)==0:
            return
        self.X[:,idx] = Z.X[:,idx]
        self.V[:,idx] = Z.V[:,idx]
        self.EX[:,idx] = Z.EX[:,idx]
        self.EV[:,idx] = Z.EV[:,idx]
        self.dEdX[:,idx] = Z.dEdX[:,idx]


class LAHMC(object):
    """
    quasiNewton Hamiltonian Monte Carlo sampler
    """

    def __init__(self, Xinit, E, dEdX, num_leapfrog_steps=10, num_look_ahead_steps=5, epsilon=0.1, alpha=0.2, beta=None, display=10):

        self.M = num_leapfrog_steps
        self.N = Xinit.shape[0]
        self.nbatch = Xinit.shape[1]
        self.display = display
        self.E_external = E
        self.dEdX_external = dEdX

        self.state = HMC_state(Xinit.copy(), self)

        self.epsilon = epsilon

        self.beta = beta
        if beta is None:
            self.beta = alpha**(1./(self.epsilon*self.M))

        # keep track of how many of each kind of transition happens
        self.counter = defaultdict(int)

    def E(self, X):
        # TODO store recent history, to avoid recalculating recent

        # TODO use parameter flattening and unflattening
        # code from SFO
        E = self.E_external(X).reshape((1,-1))
        return E

    def dEdX(self, X):
        # TODO use parameter flattening and unflattening
        # code from SFO
        dEdX = self.dEdX_external(X)
        return dEdX

    def H(self, state):
        """ The Hamiltonian for the full system """
        if state.active_idx:
            idx = state.active_idx
            H = state.EX[:,idx] + state.EV[:,idx]
        else:
            H = state.EX + state.EV
        return H

    def leapfrog(self, state):
        """ A single leapfrog step for X and V """
        if state.active_idx:
            idx = state.active_idx
            state.V[:,idx] += -self.epsilon/2. * state.dEdX[:,idx]
            state.X[:,idx] += self.epsilon * state.V[:,idx]
            state.update_dEdX()
            state.V[:,idx] += -self.epsilon/2. * state.dEdX[:,idx]
        else:
            state.V += -self.epsilon/2. * state.dEdX
            state.X += self.epsilon * state.V
            state.update_dEdX()
            state.V += -self.epsilon/2. * state.dEdX
        return state

    def L(self, state):
        """ Run the leapfrog operator for M leapfrog steps """
        for _ in range(self.M):
            state = self.leapfrog(state)
        state.update_EV()
        state.update_EX()
        return state

    def F(self, state):
        """ Flip the momentum """
        state.V = -state.V
        return state

    def leap_prob(self, Z1, Z2):
        """
        Metropolis-Hastings Probability of transitioning from state Z1 to
        state Z2.
        """
        EZ1 = self.H(Z1)
        EZ2 = self.H(Z2)
        Ediff = EZ1 - EZ2
        p_acc = np.ones(1, Ediff.shape[1])
        p_acc[Ediff<0] = np.exp(Ediff[Ediff<0])
        return p_acc

    def leap_prob_recurse(self, Z_chain, C):
        if np.isfinite(C[0,-1,0]):
            # we've already visited this leaf
            cumu = C[0,-1,:]
            return cumu, C

        if len(Z_chain) == 2:
            # the two states are one apart
            p_acc = self.leap_prob(Z_chain[0], Z_chain[1])
            C[0,-1,:] = p_acc
            return p_acc, C

        cum_forward, Cl = self.leap_prob_recurse(Z_chain[:-1], C[:-1,:-1,:])
        C[:-1,:-1,:] = Cl
        cum_reverse, Cl = self.leap_prob_recurse(Z_chain[:0:-1], C[:0:-1,:0:-1,:])
        C[:0:-1,:0:-1,:] = Cl

        H0 = self.H(Z_chain[0])
        H1 = self.H(Z_chain[1])
        start_state_ratio = np.exp(H1 - H2)
        prob = np.min(np.vstack((1. - cum_forward, start_state_ratio*(1. - cum_reverse))), 0)
        cumu = cum_forward + prob
        C[0,-1,:] = cumu
        return cumu, C

    def sampling_iteration(self):
        """ Perform a single sampling step. """

        # first do the HMC part of the step
        Z_chain = (self.state.copy(),)
        # use the same random number for comparison for the entire chain
        rand_comparison = np.random.rand(1, self.nbatch)
        # the current cumulative probability of acceptance
        p_cum = np.zeros(1, self.nbatch)
        # the cumulative probability matrix, so we only need to visit each leaf once when recursing
        C = np.ones((self.K+1, self.K+1, nbatch))*np.nan
        # the current set of indices for samples that have not yet been accepted for a transition
        active_idx = np.arange(self.nbatch, dtype=int)
        for kk in range(self.K):
            Z_chain.append(self.L(Z_chain[-1].copy()))
            # recursively calculate the cumulative probability of doing this many leaps
            p_cum, Cl = self.leap_prob_recurse(Z_chain, C[:kk, :kk, active_idx])
            # find all the samples that did this number of leaps, and update self.state with them
            accepted_idx = active_idx[p_cum >= rand_comparison[active_idx]]
            self.counter['L%d'%(kk+1)] += len(accepted_idx)
            self.state.update(accepted_idx, Z_chain[-1])
            # update the set of active indices, so we don't do simulate trajectories for samples that are already assigned to a state
            active_idx = np.flatnonzero(p_cum < rand_comparison)
            Z_chain[-1].active_idx = active_idx
        # flip the momenutm for any samples that were unable to 
        self.state.V[:,active_idx] = -self.state.V[:,active_idx]
        self.counter['F'] += len(active_idx)
        # update the current state
        self.state = SZ
        if self.display > 1:
            print "Transition counts ",
            for k in self.counter.keys():
                print "%s:%d"%(k, self.counter[k]),
            print

        # corrupt the momentum
        self.state.V = self.state.V*np.sqrt(1.-self.beta) + np.random.randn(self.N,self.nbatch)*np.sqrt(self.beta)

    def sample(self, num_steps=100):
        """
        Sample from the target distribution, for num_steps sampling steps.
        This is the function to call from external code.
        """
        for si in range(num_steps):
            if self.display > 0:
                print "sampling step %d / %d,"%(si+1, num_steps),
            self.sampling_iteration()
            if self.display > 0:
                print 
        return self.state.X

File: matlab/LAHMC.m
% Implements Look Ahead Hamiltonian Monte Carlo (LAHMC) and standard 
% Hamiltonian Monte Carlo (HMC).  See the associated paper:
%   Sohl-Dickstein, Jascha and Mudigonda, Mayur and DeWeese, Michael R.
%   Hamiltonian Monte Carlo Without Detailed Balance.
%   International Conference on Machine Learning. 2014
%
% [X, state] = LAHMC(opts, state, [additional arguments])
%
% Inputs:
%   opts - A structure setting sampling options.  Fields are:
%      Xinit - The initial particle positions for sampling.  Should be an
%        array of size [number dimensions]x[batch size].  (required field)
%      E - Function that returns the energy.  E(X, varargin{:}) should return
%        a 1x[batch size] array, with each entry corresponding to the energy
%        of the corresponding column of X.  (required field)
%      dEdX - Function that returns the gradient.  dEdX(X, varargin{:}) should
%        return an array of the same size as X,
%        [number dimensions]x[batch size].  (required field)
%      T - Number of sampling steps to take. (default 100)
%      epsilon - Step length. (default 0.1)
%      M - The number of Leapfrog steps per simulation run. (default 10)
%      K - The maximum number of times to apply the L operator. (default 4)
%      alpha - One way to specify momentum corruption rate.  See Eq. 22.
%        (default 0.2)
%      beta - Alternate way to specify momentum corruption rate.  See Eq. 15.
%        (default value derived from alpha)
%      mode - 'LAHMC' or 'HMC'. (default 'LAHMC')
%   state - The internal state of the sampler.  Set to [] the first time LAHMC
%     is called.  Pass the returned state back in to continue a sampling chain
%     across multiple calls to LAHMC.
%   [additional arguments] - Optional.  Any additional arguments are passed
%     on to the E and dEdX functions.
%     
% Outputs:
%   X - The most recent sample for every particle in the batch.
%   state - Holds the complete state of the sampler, and can be passed back in
%     to resume sampling where the last function call to LAHMC left off.
%
% See README.md for example code calling LAHMC.
%
% Author: Mayur Mudigonda, Jascha Sohl-Dickstein (2014)
% Web: http://redwood.berkeley.edu/mayur
% Web: http://redwood.berkeley.edu/jascha
% This software is made available under the Creative Commons
% Attribution-Noncommercial License.
% (http://creativecommons.org/licenses/by-nc/3.0/)


function [X, state] = LAHMC( opts, state, varargin )   
    % load parameters
    f_E = getfield(opts, 'E');
    f_dEdX = getfield(opts, 'dEdX');
    T = getField( opts, 'T', 100 );
    epsilon = getField( opts, 'epsilon', 0.1 );
    M = getField(opts, 'M', 10);
    K = getField( opts, 'K', 4);
    alpha = getField( opts, 'alpha', 0.2);
    beta = alpha^(1 / (epsilon*M));
    beta = getField( opts, 'beta', beta);    
    sampler_mode = getField( opts, 'mode', 'LAHMC');
    assert( strcmp(sampler_mode, 'HMC') | strcmp(sampler_mode, 'LAHMC') );
    
    debug_display = getField( opts, 'debug_display', 0 );

    % opts gets passed into other functions -- make these available there too
    opts.M = M; 
    opts.epsilon = epsilon;

    % initialize the state variable if not already initialized
    if isempty(state)
        state.X = getfield(opts, 'Xinit');
        szb = size(state.X, 2);
        szd = size(state.X, 1);

        state.V = randn( szd, szb );
        % state.steps provides counters for each kind of transition
        state.steps = [];
        state.steps.leap = zeros(K,1);
        state.steps.flip = 0;
        state.steps.stay = 0;
        state.steps.total = 0;
        state.funcevals = 0;
        % populate the energy and gradient at the initial position       
        state.dEdX = f_dEdX( state.X, varargin{:} );
        state.E = f_E( state.X, varargin{:} );
    end
    szb = size(state.X, 2);
    szd = size(state.X, 1);

    % used to accumulate gradient evaluations for plots in paper
    global funcevals_inc;
    funcevals_inc = 0;
    
    for t = 1:T % iterate over update steps
        if debug_display
            fprintf( '.' );
        end
        
        L_state = leap_HMC(state,[],opts,varargin{:});
        r_L = leap_prob(state,L_state,[]);
        % compare against a random number to decide whether to accept
        rnd_cmp = rand(1,szb);
        gd = (rnd_cmp < r_L);
        % update the current state for the samples where the forward transition
        % was accepted
        if sum(gd) > 0
            state = update_state(state,L_state,gd);
            state.steps.leap(1) = state.steps.leap(1) + sum(gd);
        end
        % bd indexes the samples for which the forward transition was rejected
        bd = (rnd_cmp >= r_L);
        %If there are samples that are rejected
        if sum(bd) > 0
            switch sampler_mode
                %Standard HMC flipping = 1 - leap
                case 'HMC'
                    state = flip_HMC(state,bd);
                    state.steps.flip = state.steps.flip + sum(bd);
                case 'LAHMC'
                    % n steps
                    state_ladder = {};
                    bd_idx = find(bd);
                    %bd_lad
                    state_ladder{1} = state; % Present State
                    state_ladder{2} = L_state; % Leap State

                    % this matrix will hold the cumulative probability of transitioning between pairs of states
                    C = ones(K+1,K+1,szb)*nan;

                    % Greedily evaluate how far each particle should leap, up to L^K steps
                    for nn = 3:K+1
                        state_ladder{nn} = leap_HMC(state_ladder{nn-1}, bd_idx, opts, varargin{:});
                        [~,p_cum,Cl] = leap_prob_recurse(state_ladder(1:nn), C(1:nn,1:nn,bd_idx), bd_idx);
                        C(1:nn,1:nn,bd_idx) = Cl;
                        jump_bin = rnd_cmp(bd_idx) < p_cum;
                        jump_idx = bd_idx(jump_bin);
                        state = update_state(state,state_ladder{nn},jump_idx);
                        bd_idx = bd_idx(~jump_bin);
                        state.steps.leap(nn-1) = state.steps.leap(nn-1) + length(jump_idx);
                        if length(bd_idx) == 0
                            % all the particles in the batch have found a transition
                            break
                        end
                    end
                    % and if there are any left, flip them
                    if length(bd_idx) > 0
                        state = flip_HMC(state,bd_idx);
                        state.steps.flip = state.steps.flip + length(bd_idx);
                    end
            end
        end

        % corrupt the momentum
        N1 = randn( szd, szb );      
        state.V  = real(sqrt(1-beta)) * state.V + sqrt(beta) * N1;
        state.steps.total = state.steps.total + szb;    
    end
    
    state.funcevals = state.funcevals + funcevals_inc/szb;
    
    X = state.X;
end

function [v] = getField(options,opt,default)
    % to process the fields in our options structure
    % this function taken from Mark Schmidt's minFunc
    % http://www.cs.ubc.ca/~schmidtm/Software/minFunc.html
    options = toUpper(options); % make fields case insensitive
    opt = upper(opt);
    if isfield(options,opt)
        if ~isempty(getfield(options,opt))
            v = getfield(options,opt); 
        else
            v = default;
        end
    else
        v = default;
    end
end
function [o] = toUpper(o)
    if ~isempty(o)
        fn = fieldnames(o);
        for i = 1:length(fn)
            o = setfield(o,upper(fn{i}),getfield(o,fn{i}));
        end
    end
end

function orig_state = update_state(orig_state,update_state,indices)
    % Copy the particles indicated by 'indices' from 'update_state' to 'orig_state'

    orig_state.X(:,indices) = update_state.X(:,indices);
    orig_state.V(:,indices) = update_state.V(:,indices);
    orig_state.E(:,indices) = update_state.E(:,indices);
    orig_state.dEdX(:,indices) = update_state.dEdX(:,indices);
end

function [state] = flip_HMC(state,ind)
    % apply the momentum flip operator to the particles identified by 'ind' in 'state'

    if nargin < 2
        ind = 1:size(state.V,2);
    end
    state.V(:,ind) = -state.V(:,ind);
end

function [state] = leap_HMC(state,ind,opts,varargin)
    % apply leap operator to the particles identified by 'ind' in 'state'

    global funcevals_inc

    if isempty(ind)
        ind = true(size(state.V,2),1);
    end
    Nind = sum(ind);

    f_E = opts.E;
    f_dEdX = opts.dEdX;
    % dut = 0.5; % fraction of the momentum to be replaced per unit time
    epsilon = opts.epsilon;
    M = opts.M;

    % run M leapfrog steps
    V = state.V(:,ind);
    X = state.X(:,ind);
    dEdX = state.dEdX(:,ind);

    for ii=1:M
        funcevals_inc = funcevals_inc + Nind;
        
        V0 = V;
        X0 = X;
        dEdX0 = dEdX;
        Vhalf = V0 - epsilon/2 * dEdX0;
        X1 = X0 + epsilon * Vhalf;
        dEdX1 = f_dEdX( X1, varargin{:} );
        V1 = Vhalf - epsilon/2 * dEdX1;

        V = V1;
        X = X1;
        dEdX = dEdX1;
    end
    state.X(:,ind) = X;
    state.V(:,ind) = V;
    state.dEdX(:,ind) = dEdX;
    state.E(:,ind) = f_E( X, varargin{:});
end

function [H] = hamiltonian_HMC(state,ind)
    % function to evaluate Hamiltonian of a state
    if isempty(ind)
        ind = 1:size(state.V,2);
    end
    E = state.E(:,ind);
    V = state.V(:,ind);

    % TODO store the result of (1/2) * (sum(V.*V)) so we don't have to constantly recalculate it
    H = E + (1/2) * (sum(V.*V));
end

function [resid, cumu, C] = leap_prob_recurse(state_ladder, C, idx)
    % returns [residual probability], [cumulative probability of transition from first to last state on ladder], [matrix of cumulative probabilities between all pairs of states]

    if isfinite(C(1,end,1))
        %fprintf('a')

        % this has already been calculated
        resid = 1 - C(1,end,:);
        cumu = C(1,end,:);
        resid = resid(:)';
        cumu = cumu(:)';
        return
    end

    if size(state_ladder,2) == 2
        % the two states are one apart
        prob = leap_prob(state_ladder{1}, state_ladder{2}, idx);
        cumu = prob;
        resid = 1 - prob;
        C(1,end,:) = cumu(:);
        return;
    end
       
    [residual_forward, cumulative_forward, Cl] = leap_prob_recurse(state_ladder(1:end-1), C(1:end-1,1:end-1,:), idx);
    C(1:end-1,1:end-1,:) = Cl;
    [residual_reverse, cumulative_reverse, Cl] = leap_prob_recurse(state_ladder(end:-1:2), C(end:-1:2,end:-1:2,:), idx);
    C(end:-1:2,end:-1:2,:) = Cl;

    H1 = hamiltonian_HMC(state_ladder{1},idx);
    H2 = hamiltonian_HMC(state_ladder{end}, idx);
    start_state_ratio = exp(H1 - H2);
    prob = min([residual_forward; residual_reverse.*start_state_ratio], [], 1);
    cumu = cumulative_forward + prob;
    resid = 1 - cumu;
    C(1,end,:) = cumu(:);
end

% probability of accepting a leap transition under standard
% Metropolis-Hastings
function [prob] = leap_prob(start_state, leap_state,idx)
    H_leap = hamiltonian_HMC(leap_state,idx);
    H_start = hamiltonian_HMC(start_state,idx);
    prob = min(1,exp(H_start - H_leap));
end
